src/handler/nat_hole_puncher.rs

```diff
@@ -0,0 +1,351 @@
+    pub unused_port_range: Option<RangeInclusive<u16>>,
+}
+
+impl NatHolePunchUtils {
+    pub(crate) fn new(
+        listen_port: u16,
+        local_enr: &Enr,
+        ip_mode: IpMode,
+        unused_port_range: Option<RangeInclusive<u16>>,
+    ) -> Self {
+        let mut nat_hole_puncher = NatHolePunchUtils {
+            ip_mode,
+            is_behind_nat: None,
+            new_peer_latest_relay: Default::default(),
+            hole_punch_tracker: HashSetDelay::new(Duration::from_secs(DEFAULT_HOLE_PUNCH_LIFETIME)),
+            unused_port_range,
+        };
+        // Optimistically only test one advertised socket, ipv4 has precedence. If it is
+        // reachable, assumption is made that also the other ip version socket is reachable.
+        match (
+            local_enr.ip4(),
+            local_enr.udp4(),
+            local_enr.ip6(),
+            local_enr.udp6(),
+        ) {
+            (Some(ip), port, _, _) => {
+                nat_hole_puncher.set_is_behind_nat(listen_port, Some(ip.into()), port);
+            }
+            (_, _, Some(ip6), port) => {
+                nat_hole_puncher.set_is_behind_nat(listen_port, Some(ip6.into()), port);
+            }
+            (None, Some(port), _, _) => {
+                nat_hole_puncher.set_is_behind_nat(listen_port, None, Some(port));
+            }
+            (_, _, None, Some(port)) => {
+                nat_hole_puncher.set_is_behind_nat(listen_port, None, Some(port));
+            }
+            (None, None, None, None) => {}
+        }
+        nat_hole_puncher
+    }
+
+    pub(crate) fn track(&mut self, peer_socket: SocketAddr) {
+        if self.is_behind_nat == Some(false) {
+            return;
+        }
+        self.hole_punch_tracker.insert(peer_socket);
+    }
+
+    /// Called when a new observed address is reported at start up or after a
+    /// [`crate::Discv5Event::SocketUpdated`].
+    pub(crate) fn set_is_behind_nat(
+        &mut self,
+        listen_port: u16,
+        observed_ip: Option<IpAddr>,
+        observed_port: Option<u16>,
+    ) {
+        if Some(listen_port) != observed_port {
+            self.is_behind_nat = Some(true);
+            return;
+        }
+        // Without and observed IP it is too early to conclude if the local node is behind a NAT,
+        // return.
+        let Some(ip) = observed_ip else {
+            return;
+        };
+        self.is_behind_nat = Some(is_behind_nat(ip, &self.unused_port_range));
+    }
+}
+
+impl Stream for NatHolePunchUtils {
+    type Item = Result<SocketAddr, String>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        // Until ip voting is done and an observed public address is finalised, all nodes act as
+        // if they are behind a NAT.
+        if self.is_behind_nat == Some(false) || self.hole_punch_tracker.is_empty() {
+            return Poll::Pending;
+        }
+        self.hole_punch_tracker.poll_next_unpin(cx)
```

Comment:
DelaySet can't build up indefinitely if we are not behind nat, it can only build up until we discover we are not behind nat.
https://github.com/emhane/discv5/blob/5b750584f832d44bb2fc8049295797d5fee86a18/src/handler/nat_hole_puncher.rs#L259-L264
For those first entries though, the change you propose makes sense.

Reply:
I addressed this by clearing the cache when it is confirmed this node is not behind a nat in `set_is_behind_nat` method. this means that now
`self.is_behind_nat == Some(false)` always implies `self.hole_punch_tracker.is_empty()`.